### PR TITLE
More Docker builds for Netmeld

### PR DESCRIPTION
**Pull request by cmwill**
Draft for Issue #96 
Currently untested, but I wanted to make sure the ideas are sound before I continue.

My thoughts were one image with the entire toolsuite, one image with the tools for fetching, and one image with the datalake and datastore, though the third one doesn't have psql setup yet. The playbook module is also only in the full toolsuite image and not one of its own.

I'm not yet sure on how the testing workflows could be handled. I've thought of building an image specifically for testing, but I'm unsure of it that would be enough.

### CODE REVIEW

File: docker-testing/Dockerfile-fetchers [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM netmeld-test-base
+
+ENV NETMELD_DB netmeld_test_fetchers
+
+WORKDIR /home/netmeld/setup
+
+RUN apt install --quiet --quiet --assume-yes --option Dpkg::Use-Pty=0 \
+      --no-install-recommends --no-install-suggests \
+      ./netmeld-*core.deb ./netmeld-*fetchers.deb ./netmeld-*tool-clw.deb
```
Comment (marshall-sg) [inline docker-testing/Dockerfile-fetchers:9]: Couple of things with this,

- While the `--no-install-suggests` might be ok, we want to install recommends. The use cases for not installing them are special corner cases we don't need to test under this Issue/PR.
- We don't want to explicitly call out dependencies that should already be installed prior to building this container.
- There should be a different Dockerfile for each module.

```suggestion
RUN apt install --quiet --quiet --assume-yes --option Dpkg::Use-Pty=0 \
      --no-install-suggests \
      ./netmeld-*fetchers.deb
```

File: docker-testing/Dockerfile-fetchers [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+WORKDIR /home/netmeld
+
+RUN rm -rf ./setup
+
+WORKDIR /home/netmeld/contents/fetchers
+
```
Comment (marshall-sg) [inline docker-testing/Dockerfile-fetchers:16]: Not needed as this shouldn't be the testing version.
```suggestion
```

File: docker-testing/Dockerfile-toolsuite [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+RUN apt install --quiet --quiet --assume-yes --option Dpkg::Use-Pty=0 \
+      --no-install-recommends --no-install-suggests \
+      ./netmeld-*.deb
+
+WORKDIR /home/netmeld
+
+RUN rm -rf ./setup
+
+WORKDIR /home/netmeld/contents/
+
```
Comment (marshall-sg) [inline docker-testing/Dockerfile-toolsuite:16]: Not needed as this shouldn't be the testing version.
```suggestion
```

File: docker-testing/Dockerfile-fetchers [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+CMD ["./test.py"]
```
Comment (marshall-sg) [inline docker-testing/Dockerfile-fetchers:17]: Since we aren't using for testing, it should drop to a shell or execute a command.  Furthermore, we should confine it to the `netmeld` user.  So something like:
```suggestion
USER netmeld

CMD ["/bin/bash"]
```

File: docker-testing/Dockerfile-toolsuite [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+CMD ["./toolsuite.sh"]
```
Comment (marshall-sg) [inline docker-testing/Dockerfile-toolsuite:17]: Since we aren't using for testing, it should drop to a shell or execute a command.  Furthermore, we should confine it to the `netmeld` user.  So something like:
```suggestion
USER netmeld

CMD ["/bin/bash"]
```

File: docker-testing/Dockerfile-fetchers [hunk 6]
```diff
@@ -0,0 +1,17 @@
+FROM netmeld-test-base
+
+ENV NETMELD_DB netmeld_test_fetchers
+
```
Comment (marshall-sg) [inline docker-testing/Dockerfile-fetchers:4]: Not seeing why this is here.
```suggestion
```

File: docker-testing/Dockerfile-toolsuite [hunk 7]
```diff
@@ -0,0 +1,17 @@
+FROM netmeld-test-base
+
+ENV NETMELD_DB netmeld_test_all
+
```
Comment (marshall-sg) [inline docker-testing/Dockerfile-toolsuite:4]: Not seeing why this is here.
```suggestion
```

File: docker-testing/contents/datalake-datastore/test.py [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+import subprocess
+import sys
+
+sys.path.append("..")
+from  run_command import run_command
+
+def main():
+
+    db_args = f'--db_args="host={os.environ["NETMELD_DB_HOST"]} user={os.environ["NETMELD_DB_USER"]} password={os.environ["NETMELD_DB_PASS"]}"'
+    db_name = f'--db_name="{os.environ["NETMELD_DB"]}"'
```
Comment (marshall-sg) [inline docker-testing/contents/datalake-datastore/test.py:13]: I didn't see anything in particular which you are doing that requires this to be defined.  The tools provide their own default.  For this Issue/PR, I don't think we need to support supplying custom parameters.
NOTE: This will require removal of these arguments from all following commands.
```suggestion
```

**Review by marshall-sg — CHANGES_REQUESTED** [review]
Several changes have been requested.  On top of those are the following comments:
- Some of the objects outlined in the "Describe the solution you'd like" section were missed.  I've made some comments directly in the code towards this effect.  However just to ensure, the intent is to have Docker images for each module, where it makes sense.  The testing part is an additional Docker image.  Based on what I see, it should be fairly straightforward to refactor the current code to achieve this and it may even simplify the testing logic since it only has to be done in one container versus every container.
-  Despite the ask, don't worry about integrating this into the build pipeline.  We will deal with that issue separately as we also need to be able to push them to the Docker registry for Netmeld.
